Added tutorial to submit first package to opam.ocaml.org (#18204)
ocaml/opam-repository: CONTRIBUTING.md

issue tracker to see if it doesn't concern an existing issue. If that
is the case mention it in your pull request.

Adding new packages
-------------------

Information about creating new packages and adding them to repository
is available in
[opam's manual](https://opam.ocaml.org/doc/Packaging.html).

The following is a tutorial on the steps to be followed to make your

> package available in opam.ocaml.org:

Reviewer: ```suggestion
package available in [opam.ocaml.org](https://opam.ocam.org) using `dune` as the build tool and `dune-release` as the release tool.
If your workflow does not include dune, you may try using [opam-publish](...)
```
I think we should mention that this is using a specific workflow and there are other options.

Reviewer: I think dune-release only supports github, not sure if it is the same for opam-publish. This may also need to be mentioned

Author: The objective here is to provide a `tutorial` to take the developer from point A to point B, in this case, a step-by-step procedure to get their packages submitted to opam.ocaml.org. The opam-publish instructions are already available at https://ocaml-explore.netlify.app/workflows/publishing-a-new-package-on-opam/#for-dune-users. The intent here is not to give the developer a plethora of options, but, to give them a guided approach to packaging and publishing OCaml sources.

Reviewer: I understand your perspective and I think it is good to add this.
What I wanted to say is that I think we should make extra clear that users that want an automated but less opinionated tool can also look to opam-publish, for example adding a link to the opam-publish tutorial that you just mentioned.

> A quick summary of the steps to be followed are listed below:

Reviewer: In my experience, we should also mention that in the example below we will publish an imaginary package called `hello` and that this is reflected in the name of its `opam` file in the repository. Maybe even adding something like
```
$ ls
bin/ 
src/
dune-project
hello.opam
```

issue tracker to see if it doesn't concern an existing issue. If that
is the case mention it in your pull request.

Adding new packages
-------------------

Information about creating new packages and adding them to repository
is available in
[opam's manual](https://opam.ocaml.org/doc/Packaging.html).

> The following is a tutorial that provides a step-by-step procedure to

Reviewer 2: ```suggestion
If you are using `dune` and hosting your git repository on `github` you can use the following tutorial that provides a step-by-step procedure to
```

> make a `hello` package available in opam.ocaml.org. The top-level
> sources contain the following files with the opam file having the same
> name of the project.

Reviewer 2: ```suggestion
make a `hello` package available in opam.ocaml.org.
If you are not using `dune` or hosting your project on `github` you can still use `opam-publish` as described by the manual page linked above.

The top-level sources contain the following files with the opam file having the same
name of the project.
```

```
$ ls hello/

src/
dune
dune-project
hello.opam
```

A quick summary of the steps to be followed are listed below:

```
$ dune build
$ dune-release lint
$ dune-release tag
$ dune-release distrib
$ dune-release publish
$ dune-release opam pkg

> $ dune-release opam submit

Reviewer 2: Is it worth adding (or replacing) that this can be replaced by just? Or is it worth detailing?
```
$ dune-release tag
$ dune-release
```

[-] Running package tests in _build/hello-v0.0.1
Done: 0/0 (jobs: 0)[ OK ] package tests

[+] Distribution for hello 0.0.1
[+] Commit c9935ac42271ce23cc422c299c5c3aaf92646499
[+] Archive _build/hello-v0.0.1.tbz
```

* `dune-release publish`

You can now publish the distribution to your GitHub repository as
follows:

```
$ dune-release publish
[-] Skipping documentation publication for package hello: no doc field in hello.opam
[-] Publishing distribution
[-] Publishing to github
[?] Push tag v0.0.1 to git@github.com:guest/hello? [Y/n]
Y
[-] Pushing tag v0.0.1 to git@github.com:guest/hello
[?] Create release v0.0.1 on git@github.com:guest/hello? [Y/n]
Y
[-] Creating release v0.0.1 on git@github.com:guest/hello via github's API
[+] Succesfully created release with id 38465706
[?] Upload _build/hello-v0.0.1.tbz as release asset? [Y/n]
Y
[-] Uploading _build/hello-v0.0.1.tbz as a release asset for v0.0.1 via github's API
```

* `dune-release opam pkg`

The .opam file can then be generated using the following command:

```
$ dune-release opam pkg
[-] Creating opam package description for hello
[+] Wrote opam package description _build/hello.0.0.1/opam
```

* `dune-release opam submit`

> Before you can submit the package to opam.ocaml.org, you need to first
> fork `github.com/ocaml/opam-repository` to your GitHub account, and
> also clone your copy locally to your system.
> 
> ```
> $ git clone git@github.com:guest/opam-repository
> ```
> 

Reviewer 2: ```suggestion
```
This isn't needed with my suggestion below.

> The `~/.config/dune/release.yml` file should then be updated with your
> opam repository settings as indicated below:
> 
> ```
> $ cat ~/.config/dune/release.yml
> 
> user: guest
> remote: git@github.com:guest/opam-repository
> local: /home/guest/opam-repository/

Reviewer 2: ````suggestion
The `~/.config/dune/release.yml` file should then be updated with your
opam repository settings as indicated below (in particular, change `<username>` by your local username and `<github-user>` by your own github username):

```
$ cat ~/.config/dune/release.yml

user: <username>
remote: git@github.com:<github-username>/opam-repository.git
local: /home/<username>/.cache/dune/opam-repository/
```
````
I think it is better to explicit what to change. I also changed the path of the local repository to be in a more user friendly place. See: https://github.com/ocamllabs/dune-release/issues/286
